### PR TITLE
feat(monkeys): Rewrite to allow better extensibility [WPB-3341]

### PR DESCRIPTION
**Pull request by augustocdias**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

Rewrites the internal architecture of the application in order to allow better extensibility and customisation in its execution.
Part of [WPB-3355] is also implemented here

### Issues

The application had its behaviour hardcoded.

### Solutions

New configuration options were added

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [x] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

[WPB-3355]: https://wearezeta.atlassian.net/browse/WPB-3355?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: monkeys/src/main/kotlin/com/wire/kalium/monkeys/importer/TestData.kt [hunk 1]
```diff
@@ -54,42 +54,67 @@ data class GroupConfig(
...
 
 @Serializable
-data class Action(
+data class ActionConfig(
     @SerialName("description") val description: String,
-    @SerialName("type") val type: ActionType,
+    @SerialName("config") val type: ActionType,
     @SerialName("count") val count: UserCount,
-    @SerialName("repeatDuration") val repeatDuration: UInt
+    @SerialName("repeatDuration") val repeatDuration: ULong = 0u
```
Comment (vitorhugods) [inline monkeys/src/main/kotlin/com/wire/kalium/monkeys/importer/TestData.kt:61]: What does `repeatDuration` mean? 
Is it the interval between repetitions? Maybe `repeatInterval`?

**Review by vitorhugods — APPROVED** [review]
Looking good.

Just curious about the name of a variable.

File: monkeys/src/main/kotlin/com/wire/kalium/monkeys/importer/TestData.kt (hunk 1 above)
Comment (augustocdias) [reply monkeys/src/main/kotlin/com/wire/kalium/monkeys/importer/TestData.kt:61]: yes. I'll rename it

File: monkeys/build.gradle.kts [hunk 2]
```diff
@@ -17,13 +17,12 @@
...
 
 import com.wire.kalium.plugins.commonDokkaConfig
-import com.wire.kalium.plugins.commonJvmConfig
 
 @Suppress("DSL_SCOPE_VIOLATION")
 
 plugins {
     application
-    kotlin("multiplatform")
+    kotlin("jvm")
```
Comment (typfel) [inline monkeys/build.gradle.kts:25]: Why do we need to drop Multiplatform support?

File: monkeys/src/main/kotlin/com/wire/kalium/monkeys/pool/MonkeyPool.kt [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+import java.util.concurrent.ConcurrentHashMap
+
+object MonkeyPool {
+    // a map of monkeys per domain
+    private val pool: ConcurrentHashMap<String, MutableList<Monkey>> = ConcurrentHashMap()
+
+    // a map of logged in monkeys per domain
+    private val poolLoggedIn: ConcurrentHashMap<String, ConcurrentHashMap<UserId, Monkey>> = ConcurrentHashMap()
+
+    // a map of logged in monkeys per domain
```
Comment (typfel) [inline monkeys/src/main/kotlin/com/wire/kalium/monkeys/pool/MonkeyPool.kt:32]:
```suggestion
    // a map of logged out monkeys per domain
```

**Review by typfel — CHANGES_REQUESTED** [review]
But what's the plan here you are introducing actions which can be loaded from a config file but the implementation of them is empty.

Are you planning on adding the implementation in this PR?

File: monkeys/build.gradle.kts (hunk 2 above)
Comment (augustocdias) [reply monkeys/build.gradle.kts:25]: To simplify the development. This app won't be realistically running out of the JVM

**Comment by augustocdias** [comment]
> But what's the plan here you are introducing actions which can be loaded from a config file but the implementation of them is empty.
> 
> Are you planning on adding the implementation in this PR?

The implementation of the actions will be done in a separate PR. This is to reduce the size of the PR and facilitate reviewing. This one is already too big.

**Review by typfel — APPROVED** [review]

**Pull request merged**